File: hw/PL0Compiler/Grammar Analysis/Grammar Analysis/DFA.cpp
#include"DFA.h"
#include"Trans.h"
#include"Production.h"
#include"DFAProduction.h"

namespace grammar {
	void DFA::PreProductions(const vector<Production>& grammar) {
		for (Production production : grammar) {
			//1.按产生式左部加入ref映射
			auto place = this->ref.find(production.left);

			if (place != this->ref.end()) {
				(*place).second.insert(production);
			}
			else {
				set<Production> newSet = { production };
				this->ref.emplace(production.left, newSet);
			}

			//2.给产生式编号
			this->productionToNum.emplace(production, (int)this->productionToNum.size());
			this->numToProduction.emplace((int)this->numToProduction.size(), production);
		}
	}

	void DFA::FormFirst(const Element& element) {
		if (element.ifTerminal) {
			set<string> newSet = { element.word };
			this->firstList.emplace(element.word, newSet);
		}

		else {
			this->FormFirstRecursive(Element(element.word, false));
		}
	}

	set<string> DFA::FormFirstRecursive(const Element& element) {
		if (element.ifTerminal) {  //终结符则返回只有它自己的集合
			return set<string>({ element.word });
		}

		auto place = this->firstList.find(element.word);  //非终结符如果已经求出也直接返回
		if (place != this->firstList.end()) {
			return place->second;
		}

		set<string> newSet;
		for (Production production : this->ref.find(element.word)->second) {
			vector<Element>::iterator iter = production.right.begin();  //遍历产生式右边的元素

			while (1) {
				if (iter == production.right.end()) {  //可以到达末尾，说明一路未遇到终结符，或者非终结符均有空字，将空字加入FIRST()
					newSet.insert("EPSILON");
					break;
				}

				if (iter->ifTerminal) {  //遇到终结符，加入FIRST()，并检查下一个产生式
					newSet.insert(iter->word);
					break;
				}
				
				if (iter->word == element.word) {  //遇到自己，无可用信息，检查下一个产生式
					break;
				}

				set<string> checkSet = this->FormFirstRecursive(*iter);  //对于非自己的非终结符，递归求得它的FIRST()

				if (checkSet.find("EPSILON") != checkSet.end()) {  //若含有空字，则将除空字外其余加入当前的FIRST()，并检查下一个
					checkSet.erase("EPSILON");

					for (string terminal : checkSet) {
						newSet.insert(terminal);
					}
				}

				else {  //没有空字，加入当前FIRST()后直接检查下一产生式
					for (string terminal : checkSet) {
						newSet.insert(terminal);
					}

					break;
				}

				iter++;
			}
		}

		this->firstList.emplace(element.word, newSet);  //能运行到此说明该元素的FIRST()没求出，进行记录
		return newSet;
	}

	void DFA::PreWord(const set<string>& terminals, const set<string>& nonterminals) {
		for (string terminal : terminals) {
			this->FormFirst(Element(terminal, true));  //求FIRST()
		}

		for (string nonterminal : nonterminals) {
			this->FormFirst(Element(nonterminal, false));  //求FIRST()
		}
	}

	set<string> DFA::GetFirst(const Element& word) {
		auto pair = this->firstList.find(word.word);

		if (pair == this->firstList.end()) {
			return set<string>();
		}

		return pair->second;
	}

	set<string> DFA::GetFirst(vector<Element>& wordList) {
		if (wordList.size() == 0)
			return set<string>();

		if (wordList.size() == 1)
			return this->GetFirst(wordList[0]);

		set<string> ret;
		for (Element word : wordList) {
			if (word.ifTerminal) {
				ret.insert(word.word);
				return ret;
			}

			set<string> checkSet = this->GetFirst(word);  //对于非自己的非终结符，得到它的FIRST()

			if (checkSet.find("EPSILON") != checkSet.end()) {  //若含有空字，则将除空字外其余加入当前的FIRST()，并检查下一个
				checkSet.erase("EPSILON");

				for (string terminal : checkSet) {
					ret.insert(terminal);
				}
			}

			else {  //没有空字，加入当前FIRST()后直接返回
				for (string terminal : checkSet) {
					ret.insert(terminal);
				}

				return ret;
			}
		}

		//运行到此，说明一路未遇到终结符，或者非终结符均有空字，将空字加入FIRST()
		ret.insert("EPSION");
		return ret;
	}

	void DFA::ShowGrammar() {
		for (auto pair : this->ref) {
			for (Production production : pair.second) {
				cout << production << endl;
			}
		}
		cout << endl;
	}

	void DFA::ShowFirstList() {
		for (auto pair : this->firstList) {
			cout << "FIRST(" << pair.first << ") = ";

			for (string terminal : pair.second) {
				cout << terminal << ' ';
			}

			cout << endl;
		}
	}

	void DFA::ShowDFA() {
		for (int i = 0; i < this->transList.NumOfVertices(); i++)
		{
			DFAState vertex;
			this->transList.getVertex(vertex, i);
			cout << i << ":" << endl;
			cout << vertex << endl;
		}

		for (int i = 0; i < this->transList.NumOfVertices(); i++) {
			for (int j = 0; j < this->transList.NumOfVertices(); j++) {
				string edge;

				if (this->transList.getEdge(edge, i, j)) {
					cout << i << " --" << edge << "--> " << j << endl;
				}
			}
		}
	}

	void DFA::AddTableLR1Edge(const int start, const int end, const Element& word) {
		auto transPlace = this->tableLR1[start].find(word.word);
		if (transPlace != this->tableLR1[start].end()) {  //接收该符号时，目前有结果
			transPlace->second.insert(Trans(word.ifTerminal ? TransType::ACTION_SHIFT : TransType::GOTO, end));
		}
		else {
			set<Trans> newSet;
			newSet.insert(Trans(word.ifTerminal ? TransType::ACTION_SHIFT : TransType::GOTO, end));
			this->tableLR1[start].emplace(word.word, newSet);
		}
	}

	void DFA::AddTableLR1Vertex(const DFAState& state) {
		for (DFAProduction dfaProduction : state.productions) {
			if (dfaProduction.end()) {
				Production production = Production(dfaProduction.left, dfaProduction.right);
				const int pNum = this->productionToNum.find(production)->second;

				if (dfaProduction.left == this->startWord) {  //起始产生式，填入acc
					set<Trans> newSet;
					newSet.insert(Trans(TransType::ACTION_ACC, pNum));
					this->tableLR1[state.num].emplace("TERMINAL", newSet);
				}
				else {  //其它产生式，填入r
					for (string outlook : dfaProduction.outlook) {
						auto transPlace = this->tableLR1[state.num].find(outlook);
						if (transPlace != this->tableLR1[state.num].end()) {  //接收该符号时，目前有结果
							transPlace->second.insert(Trans(TransType::ACTION_REDUCE, pNum));
						}
						else {
							set<Trans> newSet;
							newSet.insert(Trans(TransType::ACTION_REDUCE, pNum));
							this->tableLR1[state.num].emplace(outlook, newSet);
						}
					}
				}
			}
		}
	}

	DFAState& DFA::GetClosure(DFAState& state) {
		for (size_t i = 0; i < state.productions.size(); i++) {
			if (state.productions[i].end() || state.productions[i].GetPoint().ifTerminal)
				continue;

			set<string> newOutlook;
			int next = state.productions[i].point + 1;
			if (state.productions[i].nextEnd()) {
				for (string outlook : state.productions[i].outlook) {
					newOutlook.insert(outlook);
				}
			}
			else {
				vector<Element> part;

				for (size_t j = state.productions[i].point + 1; j < state.productions[i].right.size(); j++)
				{
					part.push_back(state.productions[i].right[j]);
				}

				for (string outlook : state.productions[i].outlook) {  //这个部分也要加上，即带上原产生式的展望符一起求
					part.push_back(Element(outlook, true));  //还需改进：应该是原产生式的各展望符被单带一次进行求解
				}

				newOutlook = this->GetFirst(part);
				newOutlook.erase("EPSILON");
			}

			//遍历所有*point打头的产生式，化为DFA产生式并加上求得的outlook，加入当前状态
			for (Production production : this->ref.find(state.productions[i].GetPoint().word)->second) {
				DFAProduction newDp = DFAProduction(production.left, production.right, newOutlook);
				bool ifThisAdd = true;

				for (DFAProduction dp : state.productions) {
					if (dp == newDp) {
						ifThisAdd = false;
						break;
					}
				}

				if (ifThisAdd) {
					state.productions.push_back(newDp);
				}
			}
		}

		return state;
	}

	void DFA::LR1() {
		//1.以第一个产生式构建初始状态（即outlook只有#）
		//	1.构建闭包
		//		遍历现有DFA产生式
		//		*point是终结符或已到末尾则略过
		//      *point如果是非终结符，查看*(point+1)，求FIRST()（到末尾则是当前产生式的outlook）
		//      将所有*point打头的产生式，加上求得的FIRST()集合，添加到新状态中
		//	2.构建状态迁移
		//      准备一个从当前状态转移出的状态列表（小openlist，接受符号到接受该符号后DFAState映射）
		//      遍历状态所有产生式
		//      *point到末尾则略过，否则point向后移一格后加入新state
		//		遍历完后新state加入小openlist
		//      遍历小openlist，求闭包后若已经存在，则只加边；不存在则加点，并加入openlist
		//      当前状态加入closelist
		DFAState initState;
		queue<DFAState> openlist;
		set<DFAState> closelist;

		for (Production production : this->ref.find(this->startWord)->second) {
			initState.num = 0;
			set<string> initOutLook = set<string>({ "TERMINAL" });
			initState.productions.push_back(DFAProduction(production.left, production.right, initOutLook));
		}

		this->GetClosure(initState);
		this->transList.insertVertex(initState);
		this->tableLR1.push_back(map<string, set<Trans>>());
		openlist.push(initState);
		
		while (!openlist.empty()) {
			DFAState current = openlist.front();
			openlist.pop();
			map<Element, DFAState> prepOpenlist;

			for (DFAProduction dfaProduction : current.productions) {
				if (dfaProduction.end()) {
					continue;
				}
				else {
					Element recv = dfaProduction.GetPoint();
					auto place = prepOpenlist.find(dfaProduction.GetPoint());  //取移入前，要接受的符号
					dfaProduction.add();  //点后移
					

					if (place != prepOpenlist.end()) {  //小openlist有接收该符号的状态
						place->second.productions.push_back(dfaProduction);
					}
					else {  //小openlist没有接收该符号的状态
						vector<DFAProduction> newProductions = { dfaProduction };
						DFAState newState(-1, newProductions);
						prepOpenlist.emplace(recv, newState);
					}
				}
			}

			for (auto pair : prepOpenlist) {
				Element edge = pair.first;
				DFAState state = pair.second;
				this->GetClosure(state);
				int startPos = this->transList.getVertexPos(current);
				int endPos = this->transList.getVertexPos(state);
				
				if (endPos != -1) {  //状态已经存在
					this->transList.insertEdge(startPos, endPos, edge.word);
					this->AddTableLR1Edge(startPos, endPos, edge);  //分析表添加迁移项
					if (state.num == 0)
						this->AddTableLR1Vertex(state);  //初始状态已经存在，需要分析表添加归约项
				}
				else {  //状态不存在
					state.num = this->transList.NumOfVertices();  //检查当前结点时，该结点未在图中，编号应当是顶点数
					this->transList.insertVertex(state);
					endPos = this->transList.getVertexPos(state);
					this->transList.insertEdge(startPos, endPos, edge.word);
					openlist.push(state);
					this->tableLR1.push_back(map<string, set<Trans>>());
					this->AddTableLR1Edge(startPos, endPos, edge);  //分析表添加迁移项
					this->AddTableLR1Vertex(state);                 //分析表添加归约项
				}
			}

			closelist.insert(current);
		}
	}

	string DFA::FormatTrans(set<Trans>& set) {
		string ret;
		auto iter = set.begin();

		while (iter != set.end()) {
			if (iter == set.begin()) {
				ret += iter->Format();
			}
			else {
				ret += '/' + iter->Format();
			}

			iter++;
		}

		return ret;
	}

	void DFA::ShowTableLR1(const set<string>& terminals, const set<string>& nonterminals) {
		set<int> dupStates;
		const int width = 10;

		cout << "   ";

		for (string terminal : terminals) {
			cout << setw(width) << setfill(' ') << setiosflags(ios::left) << terminal;
		}

		for (string nonterminal : nonterminals) {
			cout << setw(width) << setfill(' ') << setiosflags(ios::left) << nonterminal;
		}

		cout << endl;

		for (int i = 0; i < this->tableLR1.size(); i++) {
			DFAState state;
			this->transList.getVertex(state, i);
			cout << setw(4) << setfill(' ') << setiosflags(ios::left) << state.num;

			for (string terminal : terminals) {
				auto place = this->tableLR1[i].find(terminal);

				if (place != this->tableLR1[i].end()) {
					if (place->second.size() != 1)
						dupStates.insert(i);

					cout << setw(width) << setfill(' ') << setiosflags(ios::left) << this->FormatTrans(place->second);
				}
				else {
					for (int j = 0; j < width; j++) {
						cout << ' ';
					}
				}
			}

			for (string nonterminal : nonterminals) {
				auto place = this->tableLR1[i].find(nonterminal);

				if (place != this->tableLR1[i].end()) {
					if (place->second.size() != 1)
						dupStates.insert(i);

					cout << setw(width) << setfill(' ') << setiosflags(ios::left) << this->FormatTrans(place->second);
				}
				else {
					for (int j = 0; j < width; j++) {
						cout << ' ';
					}
				}
			}

			cout << endl;
		}

		cout << "-------------------------------" << endl;
		for (int state : dupStates) {
			cout << state << ' ';
		}
		cout << endl << "-------------------------------" << endl;
	}

	void DFA::WriteTableLR1(ofstream& out, const set<string>& terminals, const set<string>& nonterminals) {
		set<int> dupStates;
		const int width = 10;

		out << "id,";

		for (string terminal : terminals) {
			out << terminal << ',';
		}

		for (string nonterminal : nonterminals) {
			out << nonterminal << ',';
		}

		out << endl;

		for (int i = 0; i < this->tableLR1.size(); i++) {
			out << i << ',';

			for (string terminal : terminals) {
				auto place = this->tableLR1[i].find(terminal);

				if (place != this->tableLR1[i].end()) {
					if (place->second.size() != 1)
						dupStates.insert(i);

					//cout << i << "状态面临" << terminal << "操作数：" << place->second.size() << endl;
					out << this->FormatTrans(place->second) << ',';
				}
				else {
					out << ',';
				}
			}

			for (string nonterminal : nonterminals) {
				auto place = this->tableLR1[i].find(nonterminal);

				if (place != this->tableLR1[i].end()) {
					if (place->second.size() != 1)
						dupStates.insert(i);

					//cout << i << "状态面临" << nonterminal << "操作数：" << place->second.size() << endl;
					out << this->FormatTrans(place->second) << ',';
				}
				else {
					out << ',';
				}
			}

			out << endl;
		}

		out << "冲突状态,";
		for (int state : dupStates) {
			out << state << ',';
		}
	}

	vector<string> DFA::ReadWords(const char* filename) {
		ifstream in(filename);
		int count = 0;

		if (!in) {
			cout << "读取" << filename << "失败！" << endl;
			return vector<string>();
		}

		vector<string> ret;
		string word;
		in.seekg(count);
		while (in.peek() != EOF) {
			const char ch = in.peek();

			if (ch == '\n') {
				in.seekg(++count);

				if (word.length() != 0) {
					ret.push_back(word);
					word.clear();
				}
			}
			else {
				word += ch;
				in.seekg(++count);
			}
		}

		return ret;
	}

	bool DFA::GrammarAnalysis(vector<string>& input) {
		stack<int> stateRecord;
		stack<string> wordRecord;
		stateRecord.push(0);
		wordRecord.push("TERMINAL");

		set<int> ambiguousShift = {};
		set<int> ambiguousReduce = {};

		vector<string>::iterator iter = input.begin();
		while (iter != input.end()) {
			vector<string> prep;
			while (!wordRecord.empty()) {
				prep.push_back(wordRecord.top());
				wordRecord.pop();
			}
			vector<string>::iterator it = prep.end();

			do {
				it--;
				cout << *it << ' ';
				wordRecord.push(*it);
			} while (it > prep.begin());
			cout << endl;

			const int top = stateRecord.top();
			cout << top << "面临词语：" << *iter << endl;
			const Trans* op = NULL;
			auto operations = this->tableLR1[top].find(*iter);

			if (operations == this->tableLR1[top].end()) {
				cout << "throw GrammarAnalysisError" << endl;
				return false;
			}

			if (ambiguousShift.find(top) != ambiguousShift.end()) {
				for (Trans trans : operations->second) {
					if (trans.type == TransType::ACTION_SHIFT) {
						op = &trans;
					}
				}
			}
			else if (ambiguousReduce.find(top) != ambiguousReduce.end()) {
				for (Trans trans : operations->second) {
					if (trans.type == TransType::ACTION_REDUCE) {
						op = &trans;
					}
				}
			}
			else {
				op = &(*(this->tableLR1[top].find(*iter)->second.begin()));
			}

			switch (op->type) {
			case TransType::ACTION_SHIFT:
				stateRecord.push(op->dest);
				wordRecord.push(*iter);
				cout << "移进" << *iter << "迁移到" << op->dest << endl;
				break;
			case TransType::GOTO:
				stateRecord.push(op->dest);
				wordRecord.push(*iter);
				cout << "移进" << *iter << "迁移到" << op->dest << endl;
				break;
			case TransType::ACTION_REDUCE:
				{
					Production production = this->numToProduction.find(op->dest)->second;
					const string word = production.left;
					size_t reduceLen = production.right.size();

					if (reduceLen == 1 && production.right[0].word == "EPSILON")  //产生式右边只有空字时，尽管归约时栈不必弹出，但right的长度为1（存了EPSILON本身），需要设为0
						reduceLen = 0;

					for (size_t i = 0; i < reduceLen; i++) {
						stateRecord.pop();
						wordRecord.pop();
					}

					const int newTop = stateRecord.top();
					stateRecord.push(this->tableLR1[newTop].find(word)->second.begin()->dest);
					wordRecord.push(word);
					cout << "用" << op->dest << "号产生式归约到" << stateRecord.top() << endl;
					iter--;
				}
				break;
			case TransType::ACTION_ACC:
				cout << "Grammar Analysis Success" << endl;
				break;
			}

			iter++;
		}

		return true;
	}
}

File: hw/PL0_Compiler/grammar_analysis_py/DFA.py
from enum import Enum
from queue import Queue

class TransType(Enum):
    ACTION_SHIFT = 1
    ACTION_REDUCE = 2
    ACTION_ACC = 3
    GOTO = 4

class Trans:
    def __init__(self, trans_type, dest=-1):
        self.type = trans_type
        self.dest = dest

    def format(self):
        if self.type == TransType.ACTION_SHIFT:
            return f"s{self.dest}"
        elif self.type == TransType.ACTION_REDUCE:
            return f"r{self.dest}"
        elif self.type == TransType.ACTION_ACC:
            return "acc"
        elif self.type == TransType.GOTO:
            return str(self.dest)
        else:
            return ""

class DFAState:
    def __init__(self, num=-1, productions=None):
        self.num = num
        self.productions = productions or []

    def __str__(self):
        result = f"DFAState {self.num}:\n"
        result += "\n".join([str(production) for production in self.productions])
        return result

    def __repr__(self):
        return self.__str__()

from Production import Element, Production, DFAProduction

class DFA:
    def __init__(self, grammar, terminals, nonterminals, start_word):
        # 文法起始符
        self.start_word = start_word
        # str -> set<int> 产生式左部到以该词为左部产生式编号的映射
        self.production_left_ref = {}
        # Production -> num 产生式到编号的映射
        self.production_to_num = {}
        # num -> Production 包含一个列表的多个 Element
        self.num_to_production = {}
        # str -> set<str> 各个词的FIRST
        self.first_list = {}
        # list<map<str, set<Trans>>> LR1分析表
        self.table_lr1 = []
        
        self.pre_productions(grammar)
        self.pre_word(terminals, nonterminals)

    # 文法中的产生式进行预处理
    def pre_productions(self, grammar):
        num = 0

        for production in grammar:
            # 加入 产生式 - 编号 映射表
            self.production_to_num[production] = num
            self.num_to_production[num] = production

            # 左部加入
            place = self.production_left_ref.get(production.left, None)

            if place is not None:
                place.add(num)
            else:
                new_set = {num}
                self.production_left_ref[production.left] = new_set

            num += 1

    # 求取 FIRST 集合
    def pre_word(self, terminals, nonterminals):
        for terminal in terminals:
            self.form_first(Element(terminal, True))

        for nonterminal in nonterminals:
            self.form_first(Element(nonterminal, False))

    def form_first(self, element):
        if element.if_terminal:
            self.first_list[element.word] = {element.word}
        else:
            self.form_first_recursive(element)

    # 递归求非终结符的 FIRST 集合
    def form_first_recursive(self, element):
        # 终结符返回自己
        if element.if_terminal:
            return {element.word}

        # 非终结符如果已经求出也返回
        if element.word in self.first_list:
            return self.first_list[element.word]

        # 待求的 FIRST 集合
        new_set = set()

        # 以该词为左部的产生式集合
        for pro_num in self.production_left_ref[element.word]:
            
            production = self.num_to_production[pro_num]
                
            iter1 = iter(production.right)

            while True:
                try:
                    current_element = next(iter1)
                except StopIteration:
                    new_set.add("EPSILON")
                    break

                if current_element.if_terminal:
                    new_set.add(current_element.word)
                    break

                if current_element.word == element.word:
                    break

                check_set = self.form_first_recursive(current_element)

                if "EPSILON" in check_set:
                    check_set.remove("EPSILON")
                    new_set.update(check_set)
                else:
                    new_set.update(check_set)
                    break


        self.first_list[element.word] = new_set
        return new_set

    # 查询一个词的 FIRST 传入参数为 Element
    def get_first(self, word):
        pair = self.first_list.get(word.word)

        if pair is None:
            return set()

        return pair

    # 查询一个句子的 FIRST 集合 传入参数为 List<Element>
    def get_first_for_word_list(self, word_list):
        if not word_list:
            return set()

        if len(word_list) == 1:
            return self.get_first(word_list[0])

        ret = set()
        for word in word_list:
            if word.if_terminal:
                ret.add(word.word)
                return ret

            check_set = self.get_first(word)
            if "EPSILON" in check_set:
                check_set.remove("EPSILON")
                ret.update(check_set)
            else:
                ret.update(check_set)
                return ret

        ret.add("EPSILON")
        return ret

    # 打印所有 FIRST 集合
    def show_first_list(self):
        for key, value in self.first_list.items():
            print(f"FIRST({key}) = {value}")

    # 求给定 LR1 项目集的闭包 传入 DFAState
    def get_closure(self, state):
        # 遍历 state 里面的每个产生式
        i = 0
        while True:
            # print(i)
            # print(len(state.productions))
            if i >= len(state.productions):
                break

            # 规约项目 只能接收终结符
            if state.productions[i].end() or state.productions[i].get_point().if_terminal:
                i += 1
                continue

            new_outlook = set()
            next_point = state.productions[i].point + 1

            # 圆点到头
            if state.productions[i].next_end():
                new_outlook.update(state.productions[i].outlook)
            # 没到头
            else:
                for outlook in state.productions[i].outlook:
                    part = state.productions[i].right[next_point:]
                    part.append(Element(outlook, True))

                    part_first = self.get_first_for_word_list(part)
                    # new_outlook.update(part_first)
                    new_outlook = part_first.union(new_outlook)

                new_outlook.discard("EPSILON")

            left = state.productions[i].get_point().word
            for pro_num in self.production_left_ref[left]:
                production = self.num_to_production[pro_num]
                new_dp = DFAProduction(production.left, production.right, new_outlook)
                if_this_add = all(new_dp != dp for dp in state.productions)

                if if_this_add:
                    state.productions.append(new_dp)

            i += 1

    def lr1(self):
        # 初始状态
        init_state = DFAState()
        # 扩展列表
        open_list = Queue()
        # 现有状态集
        # exist_list = {}
        exist_list = []

        # Add the LR(1) project with the starting symbol as the left part of the production and the outlook symbol "TERMINAL"
        production = self.num_to_production[0]
        init_state.num = 0
        init_outlook = {"TERMINAL"}
        init_state.productions.append(DFAProduction(production.left, production.right, init_outlook))
        self.get_closure(init_state)
        # print('init state prods', init_state)

        # exist_list[tuple(init_state.productions)] = 0
        exist_list.append(init_state.productions)
        self.table_lr1.append({})  # Add a new row to the LR(1) analysis table
        open_list.put(init_state)

        cnt = 0

        while not open_list.empty():
            # if cnt > 30:
            #     break
            current = open_list.get()
            # 当前状态能转移到的所有状态集合
            prep_open_list = {}

            # 遍历当前所有产生式
            for dfa_production in current.productions:
                # 圆点到头忽略
                if dfa_production.end():
                    continue

                # 要接收的符号
                recv = dfa_production.get_point()
                # 点后移
                dfa_production.add()
                
                # print('recv', recv)
                dfa_production_copy = DFAProduction(dfa_production.left, dfa_production.right, dfa_production.outlook, dfa_production.point)

                if recv in prep_open_list:
                    # prep_open_list[recv].productions.append(dfa_production)
                    prep_open_list[recv].productions.append(dfa_production_copy)
                else:
                    # new_productions = [dfa_production]
                    new_productions = [dfa_production_copy]
                    new_state = DFAState(-1, new_productions)
                    prep_open_list[recv] = new_state

                # print('prep open list', len(prep_open_list))

            for edge, state in prep_open_list.items():
                # print('state before', state)
                self.get_closure(state)
                # print('state after', state)
                
                # print(exist_list)
                # first_key = next(iter(exist_list))
                # print(first_key)
                # print(tuple(current.productions))
                # print(first_key == tuple(current.productions))
                # print(current)
                # start_pos = exist_list[tuple(current.productions)]
                start_pos = exist_list.index(current.productions)
                # end = exist_list.get(tuple(state.productions))
                end = None
                for i, productions in enumerate(exist_list):
                    if set(productions) == set(state.productions):
                        print(productions)
                        print(state.productions)
                        end = i

                if not end:
                    print('\n\n')
                    print(exist_list)
                    print(state.productions)
                # try:
                #     end = exist_list.index(state.productions)
                #     print('\na')
                # except ValueError:
                #     end = None
                    # print('\nb',exist_list)
                # 状态在当前 DFA 中存在
                if end is not None:
                    self.add_table_lr1_edge(start_pos, end, edge)
                    if state.num == 0:
                        self.add_table_lr1_vertex(state)
                # 状态在当前 DFA 中不存在
                else:
                    state.num = len(exist_list)
                    # print("New State:", state.num)
                    # exist_list[tuple(state.productions)] = state.num
                    # print('state productions before insert', state.productions)
                    exist_list.append(state.productions)
                    open_list.put(state)
                    self.table_lr1.append({})
                    self.add_table_lr1_edge(start_pos, state.num, edge)
                    self.add_table_lr1_vertex(state)
            cnt += 1

    def add_table_lr1_edge(self, start, end, word):
        trans_place = self.table_lr1[start].get(word.word, set())

        if trans_place:
            trans_place.add(Trans(TransType.ACTION_SHIFT if word.if_terminal else TransType.GOTO, end))
        else:
            new_set = {Trans(TransType.ACTION_SHIFT if word.if_terminal else TransType.GOTO, end)}
            self.table_lr1[start][word.word] = new_set

    def add_table_lr1_vertex(self, state):
        for dfa_production in state.productions:
            if dfa_production.end():
                production = Production(dfa_production.left, dfa_production.right)
                p_num = self.production_to_num[production]

                if dfa_production.left == self.start_word:
                    new_set = {Trans(TransType.ACTION_ACC, p_num)}
                    self.table_lr1[state.num]["TERMINAL"] = new_set
                else:
                    for outlook in dfa_production.outlook:
                        trans_place = self.table_lr1[state.num].get(outlook, set())

                        if trans_place:
                            trans_place.add(Trans(TransType.ACTION_REDUCE, p_num))
                        else:
                            new_set = {Trans(TransType.ACTION_REDUCE, p_num)}
                            self.table_lr1[state.num][outlook] = new_set

    def format_trans(self, trans_set):
        ret = ""
        for trans in trans_set:
            ret += '/' + trans.format() if ret else trans.format()
        return ret

    def show_table_lr1(self, terminals, nonterminals):
        dup_states = set()
        width = 10

        print("   ", end="")
        for terminal in terminals:
            print(f"{terminal: <{width}}", end="")
        for nonterminal in nonterminals:
            print(f"{nonterminal: <{width}}", end="")
        print()

        for i in range(len(self.table_lr1)):
            print(f"{i: <4}", end="")

            for terminal in terminals:
                trans_place = self.table_lr1[i].get(terminal, set())
                if trans_place:
                    if len(trans_place) != 1:
                        dup_states.add(i)
                    print(f"{self.format_trans(trans_place): <{width}}", end="")
                else:
                    print(" " * width, end="")

            for nonterminal in nonterminals:
                trans_place = self.table_lr1[i].get(nonterminal, set())
                if trans_place:
                    if len(trans_place) != 1:
                        dup_states.add(i)
                    print(f"{self.format_trans(trans_place): <{width}}", end="")
                else:
                    print(" " * width, end="")

            print()

        print("-------------------------------")
        print(" ".join(map(str, dup_states)))
        print("-------------------------------")

def run_dfa_test():
    # Define a simple grammar
    grammar = [
        Production('S', [Element('E', False)]),
        # Production('E', [Element('E', False), Element('+', True), Element('T', False)]),
        Production('E', [Element('T', False)]),
        # Production('T', [Element('T', False), Element('*', True), Element('F', False)]),
        Production('T', [Element('F', False)]),
        Production('F', [Element('(', True), Element('E', False), Element(')', True)]),
        # Production('F', [Element('id', True)])
    ]

    terminals = {'+', '*', '(', ')', 'id'}
    nonterminals = {'S', 'E', 'T', 'F'}
    start_word = 'S'

    dfa = DFA(grammar, terminals, nonterminals, start_word)
    # dfa.show_first_list()
    dfa.lr1()

    # Show LR(1) analysis table
    # dfa.show_table_lr1(terminals, nonterminals)

if __name__ == "__main__":
    run_dfa_test()